Add untar option to dependency update

Shaps opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
Add an option to untar the dependencies after downloading. 

It's mostly a copy and resurrection of #8499 and as that, it also fixes #8479


**Special notes for your reviewer**:
I am more than happy to make more changes where required. ( I am thinking the first check box should be selected, but I guess it can be done at a later stage. ) 


**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

uvegla committed:
Delete previously successful hooks when a later hook fails

Signed-off-by: Laszlo Uveges <laszlo@giantswarm.io>

justenstall committed:
fix: replace "github.com/pkg/errors" with stdlib "errors" package

Signed-off-by: Justen Stall <39888103+justenstall@users.noreply.github.com>

justenstall committed:
more error wrapping uses

- replace os.IsNotExist with errors.Is and fs.ErrNotExist
- use %w directive

Signed-off-by: Justen Stall <39888103+justenstall@users.noreply.github.com>

justenstall committed:
restore --show-resources check in cmd/helm/status.go

Signed-off-by: Justen Stall <39888103+justenstall@users.noreply.github.com>

justenstall committed:
revert duplicate slice conditions

Signed-off-by: Justen Stall <39888103+justenstall@users.noreply.github.com>

AustinAbro321 committed:
WIP

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
extending factory to enable getting a watcher

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
basic design up and balling

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
make a working test

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

gjenkins8 committed:
refactor: Remove ChartRepository Load() function

Signed-off-by: George Jenkins <gvjenkins@gmail.com>

AustinAbro321 committed:
change structure of client

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
only emebed waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
delete wait and get completed phase

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
ability to create different waiters

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

gjenkins8 committed:
remove `ChartPaths[]`

Signed-off-by: George Jenkins <gvjenkins@gmail.com>

AustinAbro321 committed:
start watch until ready

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
lint

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
set command line flags

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
add hook only waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

AustinAbro321 committed:
cleanup

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

Shaps committed:
add untar option to dependency update

Signed-off-by: Andrea Tartaglia <me@andreatartaglia.com>

felipecrs commented on pkg/cmd/dependency_update_test.go:
```diff
@@ -234,7 +247,6 @@ func TestDependencyUpdateCmd_WithRepoThatWasNotAdded(t *testing.T) {
 		fmt.Sprintf("dependency update '%s' --repository-config %s --repository-cache %s", dir(chartname),
 			dir("repositories.yaml"), dir()),
 	)
-
```
I'd recommend undoing these cosmetic changes

Shaps replied:
yes, sorry, editor autoformat, undoing now

Shaps committed:
revert cosmetic change

Signed-off-by: Andrea Tartaglia <me@andreatartaglia.com>

AustinAbro321 committed:
legacy waiter

Signed-off-by: Austin Abro <AustinAbro321@gmail.com>

OzBenSimhonTraceloop reviewed (commented):
Any updates here guys? 🙏

nvanthao committed:
rebase

Signed-off-by: Gerard Nguyen <gerard@replicated.com>

dongjiang1989 committed:
fix codereview bug

Signed-off-by: dongjiang <dongjiang1989@126.com>

p-se committed:
Fix --take-ownership

If a resource exists in the cluster and is to be adopted by helm install
--take-ownership, it is left unchanged while helm reports the
installation to have succeeded.

This is due to CRs and CRDs being merged without three-way-merge, which
results in an empty patch.

By using a three-way-merge transparently when --take-ownership is used,
the helm behaves as expected without breaking previous behavior.

Fixes #30622

Signed-off-by: Patrick Seidensal <pseidensal@suse.com>

dongjiang1989 committed:
fix comments

Signed-off-by: dongjiang <dongjiang1989@126.com>

benoittgt committed:
Migrate to pure slog without a custom wrapper

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Call slog directly instead of using a wrapper

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

banjoh committed:
Add install test for TakeOwnership flag

Signed-off-by: Evans Mungai <mbuevans@gmail.com>

benoittgt reviewed (commented):
I’m wondering how closely we should aim to match the behavior of helm pull.

helm pull:
	•	Downloads to a temp directory when --untar is used
	•	Extracts to the final destination only after verification
	•	Automatically cleans up temp files
	•	Performs conflict detection before extraction
	•	Supports a custom untar directory via --untardir

helm dependendy update (patched):
	•	Downloads directly to the final location
	•	Leaves both the .tgz and the extracted directory (potential duplication)
	•	No conflict detection
	•	No cleanup of the original .tgz file
	•	Fixed extraction location (charts/ directory) — which I think is fine

dependabot[bot] committed:
chore(deps): bump github.com/fluxcd/cli-utils

Bumps [github.com/fluxcd/cli-utils](https://github.com/fluxcd/cli-utils) from 0.36.0-flux.13 to 0.36.0-flux.14.
- [Commits](https://github.com/fluxcd/cli-utils/compare/v0.36.0-flux.13...v0.36.0-flux.14)

---
updated-dependencies:
- dependency-name: github.com/fluxcd/cli-utils
  dependency-version: 0.36.0-flux.14
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Shaps commented on pkg/downloader/manager.go:
```diff
@@ -357,6 +358,10 @@ func (m *Manager) downloadAll(deps []*chart.Dependency) error {
 		}
 
+		if m.Untar {
+			chartutil.ExpandFile(m.ChartPath+"/charts/", tmpPath)
```
wow. Completely lost track of this PR :laughing:.
I'm not sure what you mean when you say cleaning after as suggested in the comment. I can definitely add/implement with some guidance. 
I am guessing you mean create a new tmpdir and then adding the `os.RemoveAll`, but just want to be 100% sure.

Thanks!

Shaps committed:
add verification message

`dl.DownloadTo` already handles the actual verification. This adds the
"signed by" messages if verify is defined

Signed-off-by: Andrea Tartaglia <me@andreatartaglia.com>

Shaps committed:
move saveError to collect all errors

rather than returning the latest error, collect all the errors that
happen while downloading/unarchiving and return them

Signed-off-by: Andrea Tartaglia <me@andreatartaglia.com>

Shaps committed:
extract dependencies in temp directory first

Extracts the downloaded dependencies in a temporary directory and checks
for errors before actually moving them in the final `/charts` directory

Signed-off-by: Andrea Tartaglia <me@andreatartaglia.com>

Shaps committed:
Correctly handle dependency copy

Dependencies were not being copied correctly when `--untar` was set
because when the archive is uncompressed it creates a directory
structure which was skipped by the copy code. This updates it so it
correctly handles the directories (when the untar option is set).

Additionally I found an issue when testing which prevented the archive
from being expanded because the directory was being removed before it
could be processed. This has been fixed by changing the `os.RemoveAll`
to `os.Remove` in the untar handling code. The tmp directory `tmpPath`
is already marked for deletion when it's created

Signed-off-by: Andrea Tartaglia <me@andreatartaglia.com>

Shaps commented:
I"m going to close this PR and reopen a new one with a fresh branch from main since I see the branches have diverged ( + I messed up the rebase/conflicts :laughing: ).
